builtin/logical/pki/path_acme_revoke.go

```diff
@@ -0,0 +1,153 @@
+	addFieldsForACMEPath(fields, pattern)
+	addFieldsForACMERequest(fields)
+
+	return &framework.Path{
+		Pattern: pattern,
+		Fields:  fields,
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback:                    b.acmeParsedWrapper(b.acmeRevocationHandler),
+				ForwardPerformanceSecondary: false,
+				ForwardPerformanceStandby:   true,
+			},
+		},
+
+		HelpSynopsis:    "",
+		HelpDescription: "",
+	}
+}
+
+func (b *backend) acmeRevocationHandler(acmeCtx *acmeContext, r *logical.Request, fields *framework.FieldData, userCtx *jwsCtx, data map[string]interface{}) (*logical.Response, error) {
+	var cert *x509.Certificate
+
+	rawCertificate, present := data["certificate"]
+	if present {
+		certBase64, ok := rawCertificate.([]byte)
+		if !ok {
+			return nil, fmt.Errorf("invalid type (%T) for field 'termsOfServiceAgreed': %w", rawCertificate, ErrMalformed)
+		}
+
+		certBytes, err := base64.RawURLEncoding.DecodeString(string(certBase64))
+		if err != nil {
+			return nil, fmt.Errorf("failed to base64 decode certificate: %v: %w", err, ErrMalformed)
+		}
+
+		cert, err = x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse certificate: %v: %w", err, ErrMalformed)
+		}
+	} else {
+		return nil, fmt.Errorf("bad request was lacking required field 'certificate': %w", ErrMalformed)
+	}
+
+	_, present = data["reason"]
+	if present {
+		return nil, fmt.Errorf("Vault does not support revocation reasons: %w", ErrBadRevocationReason)
+	}
+
+	if cert.NotAfter.Before(time.Now()) {
+		return nil, fmt.Errorf("refusing to revoke expired certificate: %w", ErrMalformed)
+	}
+
+	config, err := b.crlBuilder.getConfigWithUpdate(acmeCtx.sc)
+	if err != nil {
+		return nil, fmt.Errorf("unable to revoke certificate: failed reading revocation config: %v: %w", err, ErrServerInternal)
+	}
+
+	if !userCtx.Existing {
+		return b.acmeRevocationByPoP(acmeCtx, r, fields, userCtx, data, cert, config)
+	}
+
+	return b.acmeRevocationByAccount(acmeCtx, r, fields, userCtx, data, cert, config)
+}
+
+func (b *backend) acmeRevocationByPoP(acmeCtx *acmeContext, r *logical.Request, fields *framework.FieldData, userCtx *jwsCtx, data map[string]interface{}, cert *x509.Certificate, config *crlConfig) (*logical.Response, error) {
+	// Since this account does not exist, ensure we've gotten a private key
+	// matching the certificate's public key.
+	signer, ok := userCtx.Key.Key.(crypto.Signer)
+	if !ok {
+		return nil, fmt.Errorf("unable to revoke certificate: unable to parse JWS key of type (%T): %w", userCtx.Key.Key, ErrMalformed)
+	}
+
+	// Ensure that our PoP is indeed valid.
+	if err := validatePrivateKeyMatchesCert(signer, cert); err != nil {
+		return nil, fmt.Errorf("unable to revoke certificate: unable to verify proof of possession: %v: %w", err, ErrMalformed)
+	}
+
+	// Load our certificate from storage to ensure it exists.
+	serial := serialFromCert(cert)
```

Comment:
Both of these I think will need to trap and return alreadyRevoked error to the client